stage1/init: nil pointer exception when using `--private-net` and forwarded ports

When using `--private-net` and forwarded ports, `init.go` throws a nil pointer exception if the 'app' is not defined on the pod manifest json file:
```
2015/06/04 07:42:21 Execing /init
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x4109a5]

goroutine 16 [running]:
runtime.panic(0x7bf480, 0xc210f3)
	/usr/local/go/src/pkg/runtime/panic.c:279 +0xf5
main.forwardedPorts(0xc208004360, 0x0, 0x0, 0x0, 0x0, 0x0)
	/data/home/kibrahim/rkt/stage1/init/init.go:296 +0x7e5
main.stage1(0x0)
	/data/home/kibrahim/rkt/stage1/init/init.go:570 +0x6ff
main.main()
	/data/home/kibrahim/rkt/stage1/init/init.go:811 +0x4b
```

See: https://github.com/coreos/rkt/commit/6a97d6f65e0b65a084636e8db6e404c7f43c2df0


Perhaps if `a.App` is nil, we should grab it out of the `pod.Apps` map?

```
		for _, ra := range pod.Manifest.Apps {
			a := ra.App
			if a == nil {
				a = pod.Apps[ra.Name.String()].App
			}
			for _, p := range a.Ports {
				if p.Name == ep.Name {
					if n == "" {
						fp.Protocol = p.Protocol
						fp.HostPort = ep.HostPort
						fp.PodPort = p.Port
						n = ra.Name.String()
					} else {
						return nil, fmt.Errorf("Ambiguous exposed port in PodManifest: %q and %q both define port %q", n, ra.Name, p.Name)
					}
				}
			}
		}
```

--- a/stage1/init/pod.go
+++ b/stage1/init/pod.go
@@ -74,7 +74,7 @@ func LoadPod(root string, uuid *types.UUID) (*Pod, error) {
 	}
 	p.Manifest = pm
 
-	for _, app := range p.Manifest.Apps {
+	for i, app := range p.Manifest.Apps {
 		ampath := common.ImageManifestPath(p.Root, app.Image.ID)
 		buf, err := ioutil.ReadFile(ampath)
 		if err != nil {
@@ -88,6 +88,9 @@ func LoadPod(root string, uuid *types.UUID) (*Pod, error) {
 		name := am.Name.String()
 		if _, ok := p.Apps[name]; ok {
 			return nil, fmt.Errorf("got multiple definitions for app: %s", name)
+		}
+		if app.App == nil {
+			p.Manifest.Apps[i].App = am.App
 		}
 		p.Apps[name] = am
 	}
@@ -157,13 +160,10 @@ func (p *Pod) WritePrepareAppTemplate(version string) error {
 }
 
 // appToSystemd transforms the provided RuntimeApp+ImageManifest into systemd units
-func (p *Pod) appToSystemd(ra *schema.RuntimeApp, am *schema.ImageManifest, interactive bool) error {
+func (p *Pod) appToSystemd(ra *schema.RuntimeApp, interactive bool) error {
 	name := ra.Name.String()
 	id := ra.Image.ID
-	app := am.App
-	if ra.App != nil {
-		app = ra.App
-	}
+	app := ra.App
 
 	workDir := "/"
 	if app.WorkingDirectory != "" {
@@ -351,7 +351,7 @@ func (p *Pod) PodToSystemd(interactive bool) error {
 			// should never happen
 			panic("app not found in pod manifest")
 		}
-		if err := p.appToSystemd(ra, am, interactive); err != nil {
+		if err := p.appToSystemd(ra, interactive); err != nil {
 			return fmt.Errorf("failed to transform app %q into systemd service: %v", am.Name, err)
 		}
 	}
@@ -361,14 +361,11 @@ func (p *Pod) PodToSystemd(interactive bool) error {
 
 // appToNspawnArgs transforms the given app manifest, with the given associated
 // app image id, into a subset of applicable systemd-nspawn argument
-func (p *Pod) appToNspawnArgs(ra *schema.RuntimeApp, am *schema.ImageManifest) ([]string, error) {
+func (p *Pod) appToNspawnArgs(ra *schema.RuntimeApp) ([]string, error) {
 	args := []string{}
 	name := ra.Name.String()
 	id := ra.Image.ID
-	app := am.App
-	if ra.App != nil {
-		app = ra.App
-	}
+	app := ra.App
 
 	vols := make(map[types.ACName]types.Volume)
 
@@ -450,7 +447,7 @@ func (p *Pod) PodToNspawnArgs() ([]string, error) {
 		if ra == nil {
 			panic("could not find app in pod manifest!")
 		}
-		aa, err := p.appToNspawnArgs(ra, am)
+		aa, err := p.appToNspawnArgs(ra)
 		if err != nil {
 			return nil, err
 		}

--- a/stage1/init/pod_test.go
+++ b/stage1/init/pod_test.go
@@ -15,11 +15,7 @@
 package main
 
 import (
-	"strings"
 	"testing"
-
-	"github.com/coreos/rkt/Godeps/_workspace/src/github.com/appc/spec/schema"
-	"github.com/coreos/rkt/Godeps/_workspace/src/github.com/appc/spec/schema/types"
 )
 
 func TestQuoteExec(t *testing.T) {
@@ -46,83 +42,6 @@ func TestQuoteExec(t *testing.T) {
 		o := quoteExec(tt.input)
 		if o != tt.output {
 			t.Errorf("#%d: expected `%v` got `%v`", i, tt.output, o)
-		}
-	}
-}
-
-// TestAppToNspawnArgsOverridesImageManifestReadOnly tests
-// that the ImageManifest's `readOnly` volume setting will be
-// overrided by PodManifest.
-func TestAppToNspawnArgsOverridesImageManifestReadOnly(t *testing.T) {
-	falseVar := false
-	trueVar := true
-	tests := []struct {
-		imageManifestVolumeReadOnly bool
-		podManifestVolumeReadOnly   *bool
-		expectReadOnly              bool
-	}{
-		{
-			false,
-			nil,
-			false,
-		},
-		{
-			false,
-			&falseVar,
-			false,
-		},
-		{
-			false,
-			&trueVar,
-			true,
-		},
-		{
-			true,
-			nil,
-			true,
-		},
-		{
-			true,
-			&falseVar,
-			false,
-		},
-		{
-			true,
-			&trueVar,
-			true,
-		},
-	}
-
-	for i, tt := range tests {
-		imageManifest := &schema.ImageManifest{
-			App: &types.App{
-				MountPoints: []types.MountPoint{
-					{
-						Name:     "foo-mount",
-						Path:     "/app/foo",
-						ReadOnly: tt.imageManifestVolumeReadOnly,
-					},
-				},
-			},
-		}
-		podManifest := &schema.PodManifest{
-			Volumes: []types.Volume{
-				{
-					Name:     "foo-mount",
-					Kind:     "host",
-					Source:   "/host/foo",
-					ReadOnly: tt.podManifestVolumeReadOnly,
-				},
-			},
-		}
-
-		p := &Pod{Manifest: podManifest}
-		output, err := p.appToNspawnArgs(&schema.RuntimeApp{}, imageManifest)
-		if err != nil {
-			t.Errorf("#%d: unexpected error: `%v`", i, err)
-		}
-		if ro := strings.HasPrefix(output[0], "--bind-ro"); ro != tt.expectReadOnly {
-			t.Errorf("#%d: expected: readOnly: %v, saw: %v", i, tt.expectReadOnly, ro)
 		}
 	}
 }

